instrument_view.php

```diff
@@ -19,14 +19,164 @@
 }
 
 
+// Determine the request type (normal/csv/api).
+$isApiRequest = isset( $isApiRequest ) ? $isApiRequest : false;
+$isInternalRequest = isset( $isInternalRequest ) ? $isInternalRequest : false;
+$isCsvDownload = ( ! $isApiRequest && isset( $_GET['download'] ) &&
+                   $module->isReportDownloadable( $reportID ) );
+$requestType = ( ( $isApiRequest && ! $isInternalRequest )
+                 ? 'api' : ( $isCsvDownload ? 'download' : false ) );
+
+
 // Check user can view this report, redirect to main reports page if not.
-if ( ! $module->isReportAccessible( $reportID ) )
+if ( ! $isApiRequest && ! $module->isReportAccessible( $reportID ) )
 {
 	header( 'Location: ' . $module->getUrl( 'reports.php' ) );
 	exit;
 }
 
-$isCsvDownload = ( isset( $_GET['download'] ) && $module->isReportDownloadable( $reportID ) );
+
+// Define the FieldReference class used for generating editable fields.
+class FieldReference
+{
+	private $record;
+	private $fieldName;
+	private $event;
+	private $instance;
+	private $value;
+	private $label;
+
+	public function __construct( $record, $fieldName, $event, $instance, $value, $label = '' )
+	{
+		$this->record = $record;
+		$this->fieldName = $fieldName;
+		$this->event = $event;
+		$this->instance = $instance;
+		$this->value = $value;
+		$this->label = $label;
+	}
+
+	public function makeUpdateForm()
+	{
+		global $module;
+		$fieldType = \REDCap::getFieldType( $this->fieldName );
+		if ( ! in_array( $fieldType,
+		                 [ 'text', 'notes', 'radio', 'dropdown', 'yesno', 'truefalse' ] ) )
+		{
+			return $module->escapeHTML( $this->value );
+		}
+		$output = '<form method="post" class="valedit"><input type="hidden" name="record" value="' .
+		          $module->escapeHTML( $this->record ) .
+		          '"><input type="hidden" name="field" value="' .
+		          $module->escapeHTML( $this->fieldName ) . '">';
+		if ( $this->event !== null )
+		{
+			$output .= '<input type="hidden" name="event" value="' .
+			          $module->escapeHTML( $this->event ) . '">';
+		}
+		if ( $this->instance !== null )
+		{
+			$output .= '<input type="hidden" name="instance" value="' .
+			          $module->escapeHTML( $this->instance ) . '">';
+		}
+		if ( $fieldType == 'text' )
+		{
+			$output .= '<input onchange="$(this.form).submit()" type="text" name="value" value="' .
```

Comment:
I may consider this in future, but I'm concerned this may make the report look too busy.
I think the concern is mitigated by the following:
* Editable fields are not intended to be used for anything critical (e.g. trial data entry), they are intended more for entering administrative data.
* Any data entered or changed will be logged, just like entering data on a form.